Goto programs validation

sonodtt opened the pull request:
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.
--- 
2019-03-01
NB Will squash prior to merge. 
Fairly major changes after review by Kroening.
1. check for entry-point now disabled
2. checks for presence of sourcelocation and code.sourcelocation now disabled.
3. goto_modelt::validate now has default values for validation_mode and goto_model_validation_options
--- 
2019-02-27
There will be a follow up PR to this to re-enable the checks for an entry-point ("validate_goto_modelt::entry_point_exists()")
This is temporarily disabled as some (<6) regression tests have no entry point, and in some 
cases this is intentional (e.g. testing json-ui warnings)

---
2018-12-21
The first two commits are not new. Changes to reviews have been put into separate commits - but the last commit is the most consequential as it deals with @daniel.poetzl’s request for a change of location and I break the source location tests out into a new unit test file.

• Relevant to this PR (the bigger picture) are also the remarks on 
https://github.com/diffblue/cbmc/issues/751 & more recently on https://github.com/diffblue/cbmc/pull/3188
These are not addressed directly here as such, but I am flagging these as context/reminders for reviewers/ going forward.
In particular:
https://github.com/diffblue/cbmc/pull/3188#discussion_r225602406
There are several (hopefully nested but maybe not) ideas of what a "valid" goto-program is that are implicitly assumed by different parts of the code. Part of the point of all of this validity checking was to map out, clarify and correct these. Parts of the code assume that there are no RETURN instructions, parts assume that there are and things like the last instruction (apart from END_INSTRUCTION) should be return. I was asking how your validity checks will cope with this. It seems there are two options:
- The caller specifies which kind of goto-program they want to check for.
-  The check reports some auxiliary flags and the caller then checks these separately.

@peterschrammel added:
There are several kinds of Goto program that (should) form a subset chain. 
The transformations in process_Goto_program() go through this chain in several feature reduction passes. 
Certain analyses may require a specific kind of Goto program. 
So, 'valid' depends on the stage of processing. We should make this explicit.



<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,234 @@
+        const irep_idt &identifier =
+          to_symbol_expr(function_call.function()).get_identifier();
+
+        DATA_CHECK(function_map.find(identifier) != function_map.end(), "");
```
Please fill the comment (also below).

sonodtt replied:
Done. Thanks!

xbauch commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,234 @@
+    DATA_CHECK(
+      fun.second.body.instructions.back().is_end_function(),
+      "last instruction should be of end function type");
+  };
```
Is the semicolon necessary?

sonodtt replied:
erm..nope! Well spotted.

martin-cs commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -592,6 +592,14 @@ int cbmc_parse_optionst::get_goto_program(
+#if 1 // TODO remove this call location once appropriate                       \
+  // optional flags are set                                                    \
+  // (currently with '--validate-goto-model' fails due to finding              \
+  // fn pointers                                                               \
```
That statement is a little concerning.  Also, please make sure this is removed before you're done.

martin-cs commented on src/goto-programs/goto_functions.h:
```diff
@@ -121,6 +122,9 @@ class goto_functionst
   /// reported via DATA_INVARIANT violations or exceptions.
   void validate(const namespacet &ns, const validation_modet vm) const
   {
+#if 1
```
Remove before merging.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+    const validation_modet vm,
+    bool check_function_id,
+    bool check_no_fn_ptr_calls,
+    bool check_no_returns)
```
These are not the only flags you will need.  May I suggest some kind of structure so that adding flags is relatively straight forward.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+  {
+    for(auto &instr : fun.second.body.instructions)
+      DATA_CHECK(
+        instr.function == fun.first, "instruction id must be parent function");
```
Good check but you might want to check with Daniel about how long ```.function``` will exist.

sonodtt replied:
Removed this as function field will be removed.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+  bool check_no_fn_ptr_calls,
+  bool check_no_returns)
+{
+  entry_point_exists();
```
```goto-cc -c``` will generate goto-programs without entry points until they are linked so this should probably be optional.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+          to_code_function_call(instr.code);
+        DATA_CHECK(
+          function_call.function().id() == ID_symbol,
+          "no calls via fn pointer should be present");
```
s/fn/function/

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+  }
+}
+
+void validate_goto_modelt::check_called_functions()
```
Some documentation would be good.

sonodtt replied:
Added doxygen remarks above all function declarations.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+  class test_for_function_addresst : public const_expr_visitort
+  {
+  private:
+    const irep_idt match_id{ID_address_of};
```
??? Why not just use ID_address_of directly?

sonodtt replied:
Done.

martin-cs commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,237 @@
+      if(!test_for_function_address.identifiers.empty())
+      {
+        for(auto &identifier : test_for_function_address.identifiers)
+          DATA_CHECK(function_map.find(identifier) != function_map.end(),
```
Does this work for pre-linking goto-programs?

sonodtt replied:
@martin-cs 
That got us thinking... and I rewrote the following check (which now has the comments below)
  /// Check that goto-programs that have a body end with an END_FUNCTION
  /// instruction.
  ///
  /// NB functions that are only declared, and not used will not appear in the
  /// function map. Functions that are declared only and used to e.g.
  /// initialise a function pointer will have no body.
void check_last_instruction();

martin-cs reviewed (changes requested):
Right direction but not finished yet.

karkhaz commented:
@tautschnig suggested that I chime in here: `instructiont`s have a `source_location` and also a `code->source_location()`, but these are often inconsistent (sometimes one of them is `nil`). I'm wondering whether it should be an invariant that those two fields are always identical. I've been using the below code in various places, but it would be nice if there were a consistency check for this sort of thing:

```c++
const source_locationt &sloc_of(const goto_programt::const_targett &ins) const
  {
    return ins->code.source_location().is_not_nil()
             ? ins->code.source_location()
             : ins->source_location;
  }

  source_locationt &sloc_of(const goto_programt::const_targett &ins)
  {
    return const_cast<source_locationt &>(
      static_cast<const dispatch_loop_detectort &>(*this).sloc_of(ins));
  }

  const source_locationt &sloc_of(const goto_programt::instructiont &ins) const
  {
    return ins.code.source_location().is_not_nil() ? ins.code.source_location()
                                                   : ins.source_location;
  }

  source_locationt &sloc_of(const goto_programt::instructiont &ins)
  {
    return const_cast<source_locationt &>(
      static_cast<const dispatch_loop_detectort &>(*this).sloc_of(ins));
  }
```

Commit:
Validation checks for Goto Programs

There are several checks, implemented with separate methods, and
some flagged as optional.
NB use of validate_goto_model(*this, vm) in goto_functions.h is to avoid
circular inclusion as function_mapt is a nested type and so cannot be
forward declared.

sonodtt commented:
@karkhaz Many thanks for bringing this up. A check is now in place - not currently active as the code needs patching to ensure both locations are set. I'll raise an issue for that.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4f7ff243aeb4c94e5a3f0513253bad2ab25a2de0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92263816

chrisr-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+      if(!test_for_function_address.identifiers.empty())
+      {
+        for(auto &identifier : test_for_function_address.identifiers)
+          DATA_CHECK(
```
Rather than mixing up the control flow, and having to store the  identifiers to check in an intermediate set (with corresponding house keeping logic to clear the set) - would it be better to pass 'function_map' as a constructor parameter, e.g. something like `src.visit(test_for_function_addresst{function_map})` - then move this `DATA_CHECK` inside `test_for_function_addresst operator()`  - that way, you would not need to maintain an additional set, and could drop the `clear()` member function.

sonodtt replied:
Done.

chrisr-diffblue commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,62 @@
+  {
+  }
+
+  void disable_all_checks()
```
It looks like `disable_all_checks()` or `enable_all_checks()` are only used for unit tests? Even if they are, I wonder if a better pattern to follow would be to either statically declare two `goto_model_validation_optionst`s, called something like `check_all_goto_model_validation_options` and `check_no_goto_model_validation_options` (of the latter is really that useful. Alternatively, rather than providing member functions or statically declared versions - just provide a second constructor, say `goto_model_validation_optionst(bool set_all)` - which would just set all the options to the value of that bool.

sonodtt replied:
Correct, but I felt this would be useful to future users of the structure. 

Rationale was to be explicit at call location
goto_model_validation_optionst validation_options;
validation_options.disable_all_checks();
vs
goto_model_validation_optionst{false};

Have done as you suggest.

About the options struct - I think I would have preferred some iterable structure inside the options, but since that could be sliced 1000 ways - thought to start simple as you had suggested.

chrisr-diffblue commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+  return true;
+}
+
+bool test_for_failure(
```
Rather than doing this `test_for_failure` function, can you just use the built in `REQUIRE_THROWS_AS` macro whereever you otherwise use this `test_for_failure` function? e.g. this model is used in `unit/util/symbol_table.cpp`

sonodtt replied:
Done - but have kept a wrapper to keep the boilerplate to a minimum. Let me know. 
See last commit for the difference this makes.

chrisr-diffblue commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+      goto_model_validation_optionst validation_options;
+      validation_options.disable_all_checks();
+      validation_options.entry_point_exists = true;
+      REQUIRE(test_for_failure(goto_model, validation_options));
```
As mentioned above, use the built-in `REQUIRE_THROWS_AS` macro instead, ditto for other uses below.

sonodtt replied:
Done.

chrisr-diffblue reviewed (changes requested):
Overall looks good - just a few cleanup requests really.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+{
+  validate_goto_modelt{goto_functions, vm}.do_goto_program_checks(
+    validation_options);
+}
```
I think this function is unnecessary. We could instead instantiate `validate_goto_modelt` in `goto_functionst::validate()`

sonodtt replied:
Hmm, I'd rather not.
The design intent was to reduce dependencies upon validate_goto_modelt - to clarify this I have wrapped it in an unnamed namespace. 
A global function is provided - this seems a common cbmc pattern.

danpoe commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,62 @@
+  bool check_returns_removed;
+  bool check_called_functions;
+  bool check_last_instruction;
+  bool check_sourcecode_location;
```
I think the last two checks should always be performed and not be configurable. Then we could also move them out of `validate_goto_modelt` and move them into `goto_functiont::validate()` and `instructiont::validate()` respectively (then we also don't need to replicate the iteration over the goto model for those checks).

sonodtt replied:
Done. 
There were a large number of consequences to this. In particular instructiont::validate() now requires an options struct.
The source location checks are disabled atm (never called) - the discussion with Chris R. was that these could be enabled in the future. But that would require cleanup - there is ticket https://diffblue.atlassian.net/browse/ADA-215 for this future cleanup.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+
+  const validation_modet vm;
+  const function_mapt &function_map;
+  const goto_model_validation_optionst validation_options;
```
This is currently not used and should be removed.

sonodtt replied:
Done.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// 2. every instruction source_location field is nil
+  /// NB this check is not expected to pass until setting these locations is
+  /// patched up.
+  void check_sourcecode_location();
```
`sourcecode_location` -> `source_location` (also in other places)

sonodtt replied:
Done.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// until they are linked.
+  void entry_point_exists();
+
+  /// Check that no functions calls via function pointer are present
```
`pointer` -> `pointers` (also in other places)

sonodtt replied:
I'm going to assume this falls into the nitpick category and in this case I'd say that the singular case sound okay to a native speaker - it does not imply 'a single function pointer' but is a category (singular). E.g. the below with plural would not sound quite as natural. 

function_pointer_calls_removed();
"all calls via function pointer have been removed"

tautschnig replied:
s/functions/function/

sonodtt replied:
Done.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+{
+  DATA_CHECK(
+    function_map.find(goto_functionst::entry_point()) != function_map.end(),
+    "an entry point must exist");
```
Btw the `DATA_CHECK()` macros now take `vm` as an explicit parameter.

sonodtt replied:
Done. Own commit.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+          function_call.lhs().is_nil(), "function call return should be nil");
+
+        const auto &callee = to_code_type(function_call.function().type());
+        DATA_CHECK(callee.return_type().id() == ID_empty, "");
```
Add message

sonodtt replied:
Done.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+
+  /// Check both
+  /// 1. every instruction \"code\" field, has a non nil source location"
+  /// 2. every instruction source_location field is nil
```
Is non nil?

sonodtt replied:
Done.

danpoe commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+      DATA_CHECK(
+        instr.source_location.is_not_nil(),
+        "each instruction source location must not be nil");
+    }
```
Could a check be added to check that both source locations are the same?

sonodtt replied:
Done. See new unit test file "goto_program_instruction.cpp".

Commit:
Add additional 'vm' argument to DATA_CHECK

Commit:
Use unnamed namespace for validate_goto_modelt

The overall pattern here is a global function that instantiates a
utility class for a specific purpose (not a general utility class).
This clarifies that separation of concerns.
Class 'validate_goto_modelt' does not need to be globally visible.

Commit:
Improved function call check

Commit:
Use catch exception macros

Commit:
Add boolean constructor to options structure

Commit:
Add options structure to unit tests

Commit:
Adjustments for new location of program checks

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -30,32 +30,32 @@ Author: Daniel Kroening, kroening@kroening.com
+#include <goto-instrument/cover.h>
+#include <goto-instrument/full_slicer.h>
+#include <goto-instrument/nondet_static.h>
+#include <goto-instrument/reachability_slicer.h>
```
Nit pick: I'd prefer for the `goto-instrument` includes to be separated by a blank line and kept after the `goto-programs` includes. Rationale: 1) `goto-programs` is more fundamental than `goto-instrument`; 2) if there are any missing includes in some header in `goto-programs` then it must not be the case that some header in `goto-instrument` incidentally provides such an include, and thus the issue is left unnoticed.

sonodtt replied:
Noticed the implications of (2). Done.

tautschnig commented on src/goto-programs/goto_functions.h:
```diff
@@ -16,6 +16,7 @@ Date: June 2003
 
 #include "goto_function.h"
 
+#include "validate_goto_model.h"
```
Place below the include from `util` and separate by a blank line.

sonodtt replied:
Done.

tautschnig commented on src/goto-programs/goto_functions.h:
```diff
@@ -122,8 +123,12 @@ class goto_functionst
+    const validation_modet vm,
+    const goto_model_validation_optionst &goto_model_validation_options) const
   {
+    validate_goto_model(*this, vm, goto_model_validation_options);
```
Would you mind adding a comment what this is meant to do? We are inside an object that is part of any `goto_modelt`, and thus I'd expect this call to actually cause unbounded recursion?!

sonodtt replied:
Ah. No, no recursion involved. But that I should cause anyone to wonder about that is not a good thing... 

Thought to add comment 
    // validate_goto_model does a number of checks to ensure the goto_program
    // is well formed. No recursion involved as operates at the function_mapt
    // level and does not call goto_functionst::validate
    // (or any other validate methods at the goto_program level or below).

But to avoid any confusion about such recursion being possible seems best to make this call in 
```
  goto_modelt::validate (
 	validate_goto_model(
goto_functions, vm, goto_model_validation_options);
goto_functions.validate(ns, vm,  goto_model_validation_options);
)
```

ADDENDUM:

My original line of thought:

This is in the original commit:
" NB use of validate_goto_model(*this, vm) in goto_functions.h is to avoid
circular inclusion as function_mapt is a nested type and so cannot be
forward declared."

I thought seeing this in goto_modelt::validate might be confusing/debated:
```
goto_modelt::validate (
validate_goto_model(
goto_functions, vm, goto_model_validation_options);   
goto_functions.validate(ns, vm,  goto_model_validation_options);
)
```

So I chose this next-best location
```
goto_functionst::validate(
validate_goto_model(*this, vm, goto_model_validation_options);
)

```
But I think it is far far better to avoid all doubts about recursion being possible :)

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+Date: Oct 2018
+
+\*******************************************************************/
+#include "validate_goto_model.h"
```
Nit pick: add a blank line before this one.

sonodtt replied:
Done.

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+#include <set>
+
+#include "goto_functions.h"
+#include <util/invariant.h>
```
Swap the order of these includes and insert a blank line in between.

sonodtt replied:
Done.

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  using function_mapt = goto_functionst::function_mapt;
+
+  validate_goto_modelt(
+    const goto_functionst &goto_functions,
```
Why does this take `goto_functionst` when it says `validate_goto_model`?

sonodtt replied:
I passed in the highest level structure that is really needed. 
(intended as a form of documentation)

It is called goto_model as the command-line parameter is " --validate-goto-model"
I felt if I did things otherwise - that would _also_ be confusing. 

Open to suggestions.

tautschnig replied:
Ok, but I suspect you will eventually need to take a full `goto_modelt` to do some namespace lookups. But it's ok to leave it as is for now if you don't need that at the moment.

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// Check returns have been removed
+  ///
+  /// NB Calls via function pointer must have been removed already when
+  /// removing returns, thus enabling this check also should enable the check
```
The use of "should" suggests that it hasn't been done?

sonodtt replied:
Ah, yes, ambiguous - it does enable the check. 
Adjusted: "thus enabling this check also enables"

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// a corresponding entry exists in the function_map
+  void check_called_functions();
+
+  /// Check that goto-programs that have a body end with an END_FUNCTION
```
s/end/ending/

sonodtt replied:
Has been moved to goto_functiont::validate
"// function body must end with an END_FUNCTION instruction"

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// Check that for all
+  /// 1. functions that are called or
+  /// 2. functions of which the address is taken,
+  /// a corresponding entry exists in the function_map
```
Does this render properly in Doxygen? I'd be worried that we end up with "Check that for all 1" or "Check that for all" as brief description.

sonodtt replied:
Ah no - it did not render properly.  "Check that for all" was the brief description.

This fixes it:
  /// Check that for all:
  /// -# functions that are called or
  /// -# functions of which the address is taken
  /// .
  /// a corresponding entry exists in the function_map

nb the "." is required so that it shows as so:
![cprover__anonymous_namespace validate_goto_model_cpp __validate_goto_modelt_class_reference](https://user-images.githubusercontent.com/33834914/52866313-4b219d80-3136-11e9-9b49-49c4958b82e0.jpg)

a "\<br\>" or "\n" don't cut it:
(below is with "\<br\>". \n elides the last line entirely)

![cprover__anonymous_namespace validate_goto_model_cpp __validate_goto_modelt_class_reference 2](https://user-images.githubusercontent.com/33834914/52866714-2c6fd680-3137-11e9-9b64-2b765d76c964.jpg)

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  /// Check that goto-programs that have a body end with an END_FUNCTION
+  /// instruction.
+  ///
+  /// NB functions that are only declared, and not used will not appear in the
```
"will not" -> "must not" or "may or may not"?

sonodtt replied:
"do not".  These remarks were documented empirically observed behaviour.

The original comment was 
"  /// NB functions that are only declared, and not used will not appear in the
  /// function map. Functions that are declared only and used to e.g.
  /// initialise a function pointer will have no body."

This section has been moved to goto_functiont::validate, and these comments purged - as they seemed out of place in the "validate" method. Happy to add them back in if wished - any suggested location?

tautschnig replied:
At present, maybe `validate` isn't even a bad place to hold the information. One day we are going to formalise all this somewhere, for sure :-)

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  void check_last_instruction();
+
+  /// Check both
+  /// 1. every instruction \"code\" field, has a non nil source location"
```
I'm confused by the double quotes

sonodtt replied:
Removed.

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,251 @@
+  if(
+    validation_options.function_pointer_calls_removed ||
+    validation_options.check_returns_removed)
+    function_pointer_calls_removed();
```
Please use braces when the condition spans multiple lines.

sonodtt replied:
Done.

tautschnig commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,62 @@
+Date: Oct 2018
+
+\*******************************************************************/
+#ifndef CPROVER_GOTO_PROGRAMS_VALIDATE_GOTO_MODEL_H
```
Add a blank line before this one.

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+/*******************************************************************\
+
+ Module: Unit tests for goto program validation
```
Nit pick: no space before Module, Author, Date

sonodtt replied:
Done. Elsewhere too.

tautschnig replied:
Needs to be addressed again.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+code_function_callt make_void_call(const symbol_exprt &function)
+{
+  code_function_callt ret;
+  ret.function() = function;
```
There is a constructor in `code_function_callt` for this case.

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+
+  code_blockt f_body;
+  f_body.copy_to_operands(code_declt(x.symbol_expr()));
+  f_body.copy_to_operands(
```
Use the initializer-list constructor.

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+
+  code_blockt g_body;
+  g_body.copy_to_operands(code_declt(y.symbol_expr()));
+  g_body.copy_to_operands(
```
Initializer-list constructor

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+  entry_point_body.copy_to_operands(
+    code_assignt(z.symbol_expr(), from_integer(3, signed_int_type())));
+
+  entry_point_body.copy_to_operands(
```
Initializer-list constructor

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+      k.type = code_typet({}, empty_typet()); // void return, take no params
+
+      code_function_callt function_call;
+      function_call.function() = dereference_exprt(
```
Use constructor

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,429 @@
+        fn_ptr.symbol_expr(), pointer_typet(code_typet{{}, empty_typet()}, 64));
+
+      code_blockt k_body;
+      k_body.copy_to_operands(function_call);
```
Initializer-list constructor

sonodtt replied:
Done.

tautschnig commented on unit/Makefile:
```diff
@@ -60,7 +60,7 @@ SRC += analyses/ai/ai.cpp \
        util/small_map.cpp \
        util/small_shared_two_way_ptr.cpp \
-			 util/std_expr.cpp \
+       util/std_expr.cpp \
```
I'd recommend putting such commits in a separate PR, this would likely have been merged weeks ago...

sonodtt replied:
Noted. Will split this into separate PR now.

tautschnig commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -169,28 +169,40 @@ void validate_goto_modelt::check_called_functions()
+      : function_map{function_map}, vm{vm}
+    {
+    }
     void operator()(const exprt &expr) override
```
Blank lines between function definitions are free of charge ;-)

sonodtt replied:
:) Done.

tautschnig commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -31,24 +31,14 @@ struct goto_model_validation_optionst
   }
 
-  void disable_all_checks()
+  goto_model_validation_optionst(bool options_value)
```
Add `explicit`

sonodtt replied:
Done.

tautschnig commented on unit/goto-programs/goto_program_instruction.cpp:
```diff
@@ -0,0 +1,110 @@
+/*******************************************************************\
+
+  Module: Unit tests for goto_program::validate
```
Nit pick: no space before Module, Author

sonodtt replied:
Done.

tautschnig replied:
Needs to be addressed again.

tautschnig commented on unit/goto-programs/goto_program_instruction.cpp:
```diff
@@ -0,0 +1,110 @@
+
+  Author: Diffblue Ltd.
+
+ \*******************************************************************/
```
Not space before `\`, please

sonodtt replied:
Fixed.

tautschnig commented:
@sonodtt What's the status of this one? It seems to require a rebase and work on some comments, but maybe it's also subsumed by other work?

martin-cs commented:
( I am happy to re-review; it looks like lots of the issues have been resolved but maybe it makes sense to wait until this is rebased. )

sonodtt commented:
@tautschnig  @martin-cs 
Hi, was away for a few days. Sorry delay in reply. 
I had been asked to delay this until https://github.com/diffblue/cbmc/pull/3767 was merged. 

I see this has now been merged - will resume work on this.

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,50 @@
+  {
+  }
+
+  explicit goto_model_validation_optionst(bool options_value)
```
:pick: :pick: :pick: For these sorts of things, it's IMHO clearer to use tag types (`struct initialize_all_flags_to_truet{};`), enum classes (`enum class initialize_all_flagst { to_true, to_false};`) or named constructors (`goto_model_validationst goto_model_validation_optionst::with_all_flags_disabled()`).

But that's just my personal distrust of naked booleans (I basically just don't like that I had to scroll down all the way here to find out what this constructor does), which I know isn't shared by everyone here.

tautschnig replied:
Agree, please don't use Booleans without further comment. I think in this case a comment may be good enough, but otherwise the options @hannes-steffenhagen-diffblue also seem good to me.

sonodtt replied:
Addressed via nested enum class.

hannes-steffenhagen-diffblue commented on src/goto-programs/goto_model.h:
```diff
@@ -94,12 +95,20 @@ class goto_modelt : public abstract_goto_modelt
 
+    // Does a number of checks at the function_mapt level to ensure the
+    // goto_program is well formed. Does not call any validate methods
+    // (at the goto_functiont level or below)
```
:pick: Is there a way to condense this comment and turn it into a name for this function?

hannes-steffenhagen-diffblue commented on src/goto-programs/goto_model.h:
```diff
@@ -142,12 +151,20 @@ class wrapper_goto_modelt : public abstract_goto_modelt
 
+    // Does a number of checks at the function_mapt level to ensure the
+    // goto_program is well formed. Does not call any validate methods
+    // (at the goto_functiont level or below)
```
:pick: ... especially if you need to explain it twice?

sonodtt replied:
Hmm, the name is the same as the command line parameter "validate_goto_model".
I don't see how to condense this and leaving it out may cause confusion "why is this here?"
It does not harm, may help. 
see also https://github.com/diffblue/cbmc/pull/3191#discussion_r257253294

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,223 @@
+private:
+  /// Check the goto_program has an entry point
+  ///
+  /// NB goto-cc -c will generate goto-programs without entry-points
```
Super :pick: As a non-native speaker I find "NB" much harder to read than something like "note" or "note that", might be just me though

sonodtt replied:
n.b. / NB has the advantage of being universal (derives from latin), although a somewhat science / literary construct. 
Regardless, in this case it wasn't adding much, removed.

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,223 @@
+  /// .
+  /// a corresponding entry exists in the function_map
+  ///
+  /// n.b. functions that are only declared and not used will not appear in the
```
Super :pick: again: If we do use the nb, I'd prefer if it was consistently spelled as either `n.b.` or `NB`, not both

sonodtt replied:
Addressed as above.

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,223 @@
+{
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
```
:pick: explicit `const` here

sonodtt replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,223 @@
+
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
```
Also here

sonodtt replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,223 @@
+      }
+
+      // check functions of which the address is taken
+      const exprt &src{instr.code};
```
:pick: Use assignment here, something `const auto &src = static_cast<const exprt&>(instr.code);`. As is this imho looks a bit too much like a copy constructor.

sonodtt replied:
Done.

hannes-steffenhagen-diffblue commented on unit/goto-programs/goto_program_validate.cpp:
```diff
@@ -0,0 +1,359 @@
+  return true;
+}
+
+bool test_for_pass(
```
This function is a bit weird, the way you're using it you're basically having nested requires. Either make this a `void` function (and just call it instead of wrapping it in `REQUIRE`) or just inline the check within the tests themselves (may even be clearer that way, even if it's a bit more typing).

sonodtt replied:
Hmm. Rationale here was:
a) act as an interface (robust to future changes), and also demonstrate commonality 
(so people don't check in multiple lines multiple times for possible minor differences)

b) removing boiler plate : 11 vs 66 lines. Makes the core of the tests more legible.

c)  about the nested require - the intention was to preserve the familiar structure at the call site - intention was to reduce the "this is a bit weird" factor. To some extent, I'm damned if I do, damned if I don't :)    
(clarification: it'll look slightly weird either way)

Happy to remove the nested require - although I think it could be said that it serves a purpose here.

sonodtt replied:
Gone for the less surprising option - inlined all the checks.

hannes-steffenhagen-diffblue reviewed (commented):
Mostly nitpicks

Commit:
Rebase and address review comments

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e7d94afec15cfe2472c0e17017a0f9066294179b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102458564

tautschnig commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -194,7 +194,14 @@ goto_modelt initialize_goto_model(
+    goto_model_validation_optionst goto_model_validation_options{
+      goto_model_validation_optionst ::set_optionst::all_false};
+
+    // temporarily disabled until regression tests fixed
```
How long is "temporarily" and which regression tests are these?

sonodtt replied:
Added comment - will add also to PR description.

    // temporarily disabled until adjustments made for regression tests that
    // have no entry point(e.g. testing json-ui warnings when there
    // is no entry point)

Will be the next PR. I'll start working on it right away.

kroening replied:
goto models really aren't required to have an entry point; we do want to support goto-cc -c in the long term.

tautschnig commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,69 @@
+  bool check_called_functions;
+
+  // this option is always set to false until all source locations
+  // are reliably set correctly (future work)
```
I think the comment below is sufficient, no need to single it out here as well.

sonodtt replied:
Done.

tautschnig commented on unit/Makefile:
```diff
@@ -22,9 +22,11 @@ SRC += analyses/ai/ai.cpp \
        goto-programs/goto_program_symbol_type_table_consistency.cpp \
        goto-programs/goto_program_table_consistency.cpp \
        goto-programs/goto_trace_output.cpp \
+       goto-programs/goto_program_validate.cpp \
```
Should be moved up by one line to maintain sorting.

sonodtt replied:
Ach! Done.

tautschnig commented on unit/goto-programs/goto_program_goto_target.cpp:
```diff
@@ -1,16 +1,17 @@
 /*******************************************************************\
 
-Module: Unit tests for goto_program::validate
+
```
Not an obvious improvement?

sonodtt replied:
Fixed.

tautschnig reviewed (approved):
Re-reviews from @martin-cs @chrisr-diffblue are required as well.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1fb11297d5fdce7875f7fdf3141bb531c9045ec2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102487466

kroening commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -550,7 +551,10 @@ int cbmc_parse_optionst::doit()
+    goto_model.validate(
+      validation_modet::INVARIANT,
+      goto_model_validation_optionst{
+        goto_model_validation_optionst::set_optionst::all_true});
```
Given how often that all_true option is used, simply make that the default?

kroening replied:
Same for INVARIANT?

sonodtt replied:
Done. Also for INVARIANT.

kroening commented on src/goto-programs/goto_program.cpp:
```diff
@@ -776,6 +777,27 @@ void goto_programt::instructiont::validate(
+      vm,
+      code.source_location().is_not_nil(),
+      "each instruction code field, must have non nil source location");
+
```
Remove the comma

kroening replied:
Note that not every instruction has a `code` member, so this will fail.

sonodtt replied:
Action taken (after team chat). In commit message (reproduced here in case it helps):

Removed previously disabled checks :
Not every instruction has a code member - so removed checks that both 
instruction sourcelocation and code sourcelocation are set and identical
Remove also remaining check that every instruction have a non-nil 
sourcelocation as this would have to be optional (if enabled fails 
many regression tests). This also simplifies considerably the overall 
validation pass (removes much passing around of the options structure).

kroening commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,222 @@
+  /// - lhs of every \ref code_function_callt instruction is nil
+  /// - all return types are void (of both calls and functions themselves)
+  void check_returns_removed();
+
```
Please note #4266 on the return type.

sonodtt replied:
removed in validate_goto_model.cpp
```
    DATA_CHECK(
      vm,
      goto_function.type.return_type().id() == ID_empty,
      "functions must have empty return type");
```

kroening commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,222 @@
+      {
+        const code_function_callt &function_call =
+          to_code_function_call(instr.code);
+        DATA_CHECK(
```
Use `instr.get_function_call()`.

sonodtt replied:
Done.

kroening commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,222 @@
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        DATA_CHECK(
```
Same here

sonodtt replied:
Done.

kroening commented on src/goto-programs/validate_goto_model.cpp:
```diff
@@ -0,0 +1,222 @@
+      // check functions that are called
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
```
Same here

sonodtt replied:
Done.

kroening commented on unit/goto-programs/goto_program_instruction.cpp:
```diff
@@ -0,0 +1,116 @@
+    auto &instructions = goto_function.body.instructions;
+
+    goto_programt::instructiont instruction;
+    instruction.make_end_function();
```
instructiont::make_end_function is deprecated -- use goto_programt::make_end_function

sonodtt replied:
Done.

kroening commented on unit/goto-programs/goto_program_instruction.cpp:
```diff
@@ -0,0 +1,116 @@
+    goto_programt::instructiont instruction;
+    instruction.make_end_function();
+    instructions.push_back(instruction);
+    instructions.back().source_location.id("id_any_valid_id");
```
The source_locationt ireps don't have an ID.

sonodtt replied:
Done.

kroening commented on unit/goto-programs/goto_program_symbol_type_table_consistency.cpp:
```diff
@@ -36,6 +37,16 @@ SCENARIO(
+    // required as goto_function.validate checks (if a function has a body) that
+    // the last instruction of a function body marks the function's end.
+    goto_programt::instructiont end_function_instruction;
+    end_function_instruction.make_end_function();
```
Same here

sonodtt replied:
Done.

Commit:
Address review comments - Kroening

Removed previously disabled checks:
Not every instruction has a code member - so removed checks that both 
instruction sourcelocation and code sourcelocation are set and identical
Remove also remaining check that every instruction have a non-nil 
sourcelocation as this would have to be optional (if enabled fails 
many regression tests). This also simplifies considerably the overall 
validation pass (removes much passing around of the options structure).

Removes check on function return type - this will be preserved (#4266)

goto_model::validate now has default parameters.

Minor fixes.

Commit:
Remove extraneous goto-program checks in unit tests.

Although additional checks may hold, they have their own unit
tests and it is these tests that should fail when these checks fail.

Commit:
Reorder unit test headers for consistency.

Commit:
Address review comments - Kroening

Removed previously disabled checks:
Not every instruction has a code member - so removed checks that both 
instruction sourcelocation and code sourcelocation are set and identical
Remove also remaining check that every instruction have a non-nil 
sourcelocation as this would have to be optional (if enabled fails 
many regression tests). This also simplifies considerably the overall 
validation pass (removes much passing around of the options structure).

Removes check on function return type - this will be preserved (#4266)

goto_model::validate now has default parameters.

Minor fixes.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6ba2cf43b4d9c634fea4fe0ed6ccb48753a867b8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102836757

martin-cs commented on src/goto-programs/validate_goto_model.h:
```diff
@@ -0,0 +1,65 @@
+  bool function_pointer_calls_removed = true;
+  bool check_returns_removed = true;
+  bool check_called_functions = true;
+
```
If this is intended to be comprehensive right now, it should also check for `vector` and `complex`.  If this is intended to be a general framework then it's not super urgent.

chrisr-diffblue replied:
Its a general framework right now - everyone is invited to fill in the additional checks :-)

martin-cs reviewed (approved):
As a start this is reasonable but there are many invariants it doesn't check (whether variables follow the DECL - use - DEAD cycle, scoping of variables, whether exprt's are well formed, etc.).  I presume these will be in follow-up patches.
